monetio/sat/_tempo_l2_no2_mm.py

```diff
@@ -0,0 +1,233 @@
+
+    dso = Dataset(fname, "r")
+    lon_var = dso.groups["geolocation"]["longitude"]
+    lat_var = dso.groups["geolocation"]["latitude"]
+    time_var = dso.groups["geolocation"]["time"]
+
+    ds["lon"] = (
+        ("x", "y"),
+        lon_var[:].squeeze(),
+        {"long_name": lon_var.long_name, "units": lon_var.units},
+    )
+    ds["lat"] = (
+        ("x", "y"),
+        lat_var[:].squeeze(),
+        {"long_name": lat_var.long_name, "units": lat_var.units},
+    )
+    ds["time"] = (
+        ("time",),
+        time_var[:].squeeze(),
+        {"long_name": time_var.long_name, "units": time_var.units},
+    )
+    ds["lon"] = ds["lon"].fillna(-999.99)
+    ds["lat"] = ds["lat"].fillna(-999.99)
+    ds = ds.set_coords(["time", "lon", "lat"])
+
+    ds.attrs["reference_time_string"] = dso.time_coverage_start
+    ds.attrs["granule_number"] = dso.granule_num
+    ds.attrs["scan_num"] = dso.scan_num
+
+    for varname in variable_dict:
+        if varname in [
+            "main_data_quality_flag",
+            "vertical_column_troposphere",
+            "vertical_column_stratosphere",
+            "vertical_column_troposphere_uncertainty",
+        ]:
+            values_var = dso.groups["product"][varname]
+        elif varname in [
+            "latitude_bounds",
+            "longitude_bounds",
+            "solar_zenith_angle",
+            "solar_azimuth_angle",
+            "viewing_zenith_angle",
+            "viewing_azimuth_angle",
+            "relative_azimuth_angle",
+        ]:
+            values_var = dso.groups["geolocation"][varname]
+        elif varname in [
+            "vertical_column_total",
+            "vertical_column_total_uncertainty",
+            "fitted_slant_column",
+            "fitted_slant_column_uncertainty",
+            "snow_ice_fraction",
+            "terrain_height",
+            "ground_pixel_quality_flag",
+            "surface_pressure",
+            "tropopause_pressure",
+            "scattering_weights",
+            "gas_profile",
+            "albedo",
+            "temperature_profile",
+            "amf_total",
+            "amf_diagnositc_flag",
+            "eff_cloud_fraction",
+            "amf_cloud_fraction",
+            "amf_cloud_pressure",
+            "amf_troposphere",
+            "amf_stratosphere",
+        ]:
+            values_var = dso.groups["support_data"][varname]
+        elif varname in ["fit_rms_residual", "fit_convergence_flag"]:
+            values_var = dso.groups["qa_statistics"][varname]
+        values = values_var[:].squeeze()
+        fv = values_var.getncattr("_FillValue")
+        if not np.isfinite(fv):
+            values[:][values[:] == fv] = np.nan
+
+        if "scale" in variable_dict[varname]:
+            values[:] = variable_dict[varname]["scale"] * values[:]
+
+        if "minimum" in variable_dict[varname]:
+            minimum = variable_dict[varname]["minimum"]
+            values[:] = np.where(values[:] >= minimum, values[:], np.nan)
```

Comment:
I hadn't realized the problem of minimum being NaN, and was addressing the case of:
a) minimum not being there (in which case the np.where() function is not applied),
b) minimum being a number, in which case this works.

I'll add a check for minimum being NaN to avoid this.

The problem before was when minimum was a number (either int or float) and some values were NaN: the code would just break saying that it could not do the comparison, at least with Python 3.11.

Reply:
> The problem before was when minimum was a number (either int or float) and some values were NaN: the code would just break saying that it could not do the comparison, at least with Python 3.11.

I tried to replicate this with Python 3.11 / NumPy 2.1.0 and was not able to. Is this not the correct situation?
```python
>>> a = np.r_[1.0, 2.0, np.nan, 4.0]
>>> a[a < 2] = np.nan
>>> a
array([nan,  2., nan,  4.])
```

Reply:
OK, this is weird. I only get the bug I mentioned with the "terrain height" variable, and I don't know why.
Here is the full Traceback in an Ipython session.
It might be related to the datatype of terrain height. It is likely that that variable is not really needed, but I would rather have it as an option (since a user might want to know the height of the terrain at a certain location according to TEMPO).

```
Cell In[2], line 1
----> 1 vals = mio.sat._tempo_l2_no2_mm.open_dataset(
      2     "/glade/campaign/acom/acom-da/sma/TEMPO_NO2_V03/TEMPO_NO2_L2_V03_20240707*S011G02*.nc",
      3     {
      4         "vertical_column_stratosphere": {"minimum":0},
      5         "scattering_weights": {},
      6         "vertical_column_troposphere": {},
      7         "main_data_quality_flag": {"quality_flag_max": 0},
      8         #"surface_pressure": {"maximum": 100000},
      9         "terrain_height": {"minimum": 0}
     10     },
     11 )

File ~/melodies-monet/monetio/monetio/sat/_tempo_l2_no2_mm.py:230, in open_dataset(fnames, variable_dict, debug)
    227 granules = OrderedDict()
    229 for file in files:
--> 230     granule = _open_one_dataset(file, variable_dict)
    231     apply_quality_flag(granule)
    232     granules[granule.attrs["reference_time_string"]] = granule

File ~/melodies-monet/monetio/monetio/sat/_tempo_l2_no2_mm.py:118, in _open_one_dataset(fname, variable_dict)
    116     minimum = variable_dict[varname]["minimum"]
    117     #values[:] = np.where(values[:] >= minimum, values[:], np.nan)
--> 118     values[values < minimum] = np.nan
    120 if "maximum" in variable_dict[varname]:
    121     maximum = variable_dict[varname]["maximum"]

File /glade/work/plichtig/conda-envs/melodies-monet-test/lib/python3.11/contextlib.py:81, in ContextDecorator.__call__.<locals>.inner(*args, **kwds)
     78 @wraps(func)
     79 def inner(*args, **kwds):
     80     with self._recreate_cm():
---> 81         return func(*args, **kwds)

File /glade/work/plichtig/conda-envs/melodies-monet-test/lib/python3.11/site-packages/numpy/ma/core.py:3387, in MaskedArray.__setitem__(self, indx, value)
   3384     mval = tuple([False] * len(_dtype.names))
   3385 if _mask is nomask:
   3386     # Set the data, then the mask
-> 3387     _data[indx] = dval
   3388     if mval is not nomask:
   3389         _mask = self._mask = make_mask_none(self.shape, _dtype)

ValueError: cannot convert float NaN to integer
```

Reply:
Weird that terrain height would be int. But yes I see the issue. Easiest solution would be to convert the array to float first (`.astype(float)`) before NaN application. But it looks this may actually be a masked array, so you'd really want to modify the mask, not set values to NaN.

Thanks for working on this, which relates to #184 as well.

Reply:
Casting it is definitely a solution, but another (possibly simpler) solution is just inverting the condition of my previous code. You can look at my next commit for that. I. e., it is doing


values = np.where(values [:] < min, np.nan, values[:])

Since the condition is always false if min==np.nan , this would avoid the previous issue, and wouldn't require casting the array into float.

Although, of course, this is still modifying the values instead of the mask. I should probably take your advice and look at how to modify the mask.
terrain_height is of type "short" (ie, float32), so I'm not sure why it fails as if it were int, but it is the exact same error message. Maybe it's a "feature" of masked arrays?

Reply:
Done. Now it deals with "maximum" and "minimum" by masking the arrays. I don't know if this really makes a difference at the xarray.DataArray level.